### PR TITLE
Align the guide with the user guide chapter

### PR DESCRIPTION
**Pull request by lacasseio**
This PR must be sync with https://github.com/gradle/gradle/pull/3450 in the context of https://github.com/gradle/gradle/issues/2629.

### CODE REVIEW

File: contents/index.adoc [hunk 1]
```diff
@@ -126,7 +126,7 @@ The following example applies a plugin that exposes an extension with the name `
 include::{samplescodedir}/capture-user-input/build.gradle[]
 ----
 
-Let's assume that you'll also want to do something with the value of `serverUrl` once captured. In many cases the exposed extension property is directly mapped to a task property that actually uses the value when performing work. To avoid evaluation order problems you should use {user-manual}custom_plugins.html#mapExtensionPropertiesToTaskProperties[the public API `Property`] which was introduced in Gradle 4.0.
+Let's assume that you'll also want to do something with the value of `serverUrl` once captured. In many cases the exposed extension property is directly mapped to a task property that actually uses the value when performing work. To avoid evaluation order problems you should use {user-manual}lazy_configuration.html#lazy_properties[the public API `Property`] which was introduced in Gradle 4.0.
```
Comment (bmuschko) [inline contents/index.adoc]: Is the https://docs.gradle.org/4.3.1/userguide/lazy_configuration.html#lazy_properties going to be release with 4.4? Right now it is failing the build: https://travis-ci.org/gradle-guides/implementing-gradle-plugins/builds/300339498?utm_source=github_status&utm_medium=notification though I'd understand if it hasn't been released yet. It's a shortcoming of our core vs. guide release strategy.

File: samples/code/capture-user-input/build.gradle [hunk 2]
```diff
@@ -1,5 +1,5 @@
 apply plugin: BinaryRepositoryVersionPlugin
 
 binaryRepo {
-    serverUrl = 'http://my.company.com/maven2'
-}
\ No newline at end of file
+    serverUrl.set 'http://my.company.com/maven2'
```
Comment (bmuschko) [inline samples/code/capture-user-input/build.gradle]: As I user I wouldn't consider this usage very DSLish. Is this how we are expecting users to define `Provider` values? I believe this is the discussion we had a couple of weeks ago on Slack. If that's the general way forward then that's good. Still from a user's perspective it feels less like a DSL.

Comment (bmuschko) [reply samples/code/capture-user-input/build.gradle]: If this is the way forward then we should also make changes to https://github.com/gradle-guides/gradle-site-plugin.

File: contents/index.adoc (hunk 1 above)
Comment (lacasseio) [reply contents/index.adoc]: The plan is to release something for 4.4. Hopefully, this won't impact too much the release schedule. If we fear it's not ready yet, we can pull back and push it to 4.5

File: samples/code/capture-user-input/build.gradle (hunk 2 above)
Comment (lacasseio) [reply samples/code/capture-user-input/build.gradle]: You have a very good point. There's actually some Groovy DSL decoration that allows the DSL to stay the same as it was previously. I reverted back the change.

Comment (lacasseio) [reply samples/code/capture-user-input/build.gradle]: I updated the `gradle-site-plugin` project: https://github.com/gradle-guides/gradle-site-plugin/pull/2

Comment (bmuschko) [reply samples/code/capture-user-input/build.gradle]: We should explain how the decoration works. I don't see it mentioned in the text. That'll feel somewhat magical to end users. Do we generate the setter method for any field of type `Provider`?

Comment (lacasseio) [reply samples/code/capture-user-input/build.gradle]: I added a note to explain injected setters: https://github.com/gradle-guides/implementing-gradle-plugins/pull/16/commits/7af77dc75516ad69d5deba5148dd7e5a63e7b440

File: contents/index.adoc (hunk 1 above)
Comment (big-guy) [reply contents/index.adoc]: Ben suggested that we use a 4.4-nightly to test these changes and switch to 4.4 once it's final.

Comment (lacasseio) [reply contents/index.adoc]: Make sense, let me change that now.

Comment (bmuschko) [reply contents/index.adoc]: I guess the build is still failing for two reasons:

1. We do not publish docs for nightlies or release snapshots.
2. The new content hasn't been integrated yet.

Let's wait for RC1 is out and make the change then.

File: contents/index.adoc [hunk 3]
```diff
@@ -138,6 +138,8 @@ include::{samplescodedir}/capture-user-input/buildSrc/src/main/java/BinaryReposi
 
 Instead of using a plain `String` type, the extension defines the field `serverUrl` with type `Property<String>`. The field is initialized in the constructor of the class. It's state can be set directly on the object by getting the instance via the exposed getter methods.
 
+NOTE: Gradle classloader injects setter methods side-by-side with the all `Property` type getter methods. It allows developers to simplify code like `obj.prop.set 'foo'` to `obj.prop = 'foo'` in the Groovy DSL.
```
Comment (bmuschko) [inline contents/index.adoc]: I'd probably change the first sentence a bit: "The Gradle classloader automatically injects setter methods alongside all getter methods with the return type `Property`.".

- What currently happens for the deprecated `PropertyState`?
- What happens when using the feature with the Kotlin DSL?

Comment (lacasseio) [reply contents/index.adoc]: > What currently happens for the deprecated `PropertyState`?

I'm not sure if we should document what is happening with `PropertyState` as it's `Incubating`. I think the mention in the release note is good enough.

> What happens when using the feature with the Kotlin DSL?

I'm not sure about this one but I think it's a non-issue for Kotlin as it already provides the right keywords, see [this comment from Paul](https://github.com/gradle/tech-sessions/issues/14#issuecomment-334685622).

> Do we explain those things in the user manual? 

We don't explain specifically Kotlin DSL. There's no mention about PropertyState.

Comment (bmuschko) [reply contents/index.adoc]: > We don't explain specifically Kotlin DSL. There's no mention about PropertyState.

The plan is to explain Groovy and Kotlin DSL along side. I think more and more users will ask about it now that Kotlin has become so popular in the Android community. We should at least keep that in mind when writing content.

Comment (lacasseio) [reply contents/index.adoc]: There's a link back to the user guide at line 129.

**Pull request merged**